dbus/src/message/matchrule.rs

```diff
@@ -167,4 +167,300 @@ impl<'a> MatchRule<'a> {
+
+        /// Tokenizes a string into key-value-pairs
+        pub fn tokenize(&mut self) -> Result<TokenBuffer<'a>, Error> {
+            let mut rules = TokenBuffer::new();
+
+            while !self.text.is_empty() {
+                let (key, rest) = self.key();
+                self.text = rest;
+                let (value, rest) = self.value();
+                self.text = rest;
+                rules.try_push((key, value))
+                    .map_err(|_| Error::TooLong)?;
+            }
+            Ok(rules)
+        }
+    }
+
+    #[derive(Clone, Debug)]
+    /// Helper struct for parsing MatchRule's
+    pub struct Parser<'a> {
+        tokens: TokenBuffer<'a>,
+    }
+
+    impl<'a> Parser<'a> {
+        /// Builds a new parser after tokenizing the input string `text`.
+        pub fn new(text: &'a str) -> Result<Self, Error> {
+            Ok(Self {
+                tokens: Tokenizer::new(text).tokenize()?
+            })
+        }
+
+        /// Cleans a string from the string syntax allowed by DBus. This includes concatenating
+        /// things like `''\'''` to `'`. DBus strings sort of work like in a POSIX shell and
+        /// concatenation is implied. There is only one escape sequence and that is in an unquoted
+        /// substring a backslash may escape an ASCII apostrophe (U+0027).
+        /// This method is the only one within the parser that allocates and in theory could be
+        /// rewritten to taking a `&mut str` instead of returning a `String` since all
+        /// strings are |output| <= |buf.len()|.
+        fn clean_string(&self, buf: &str) -> String {
+            let mut quoted = false;
+            let mut escape = false;
+            let mut outbuf = String::with_capacity(buf.len());
+
+            for c in buf.chars() {
+                match c {
+                    '\'' if !escape => {
+                        quoted = !quoted;
+                    }
+                    '\\' if !quoted => {
+                        escape = true;
+                        continue;
+                    }
+                    c if c.is_whitespace() && !quoted => {
+                        continue;
+                    }
+                    c => {
+                        outbuf.push(c);
+                    }
+                }
+            }
+
+            outbuf
+        }
+
+        /// Parses key-value-pair tokens into a MatchRule
+        pub fn parse(&self) -> Result<MatchRule<'a>, Error> {
+            let mut match_rule = MatchRule::new();
+
+            for &(key, raw_value) in &self.tokens {
+                let value = self.clean_string(raw_value);
+                match key {
+                    "type" => {
+                        match_rule = match_rule.with_type(MessageType::try_from(value.as_str())?);
+                        Ok(())
+                    }
+                    "interface" => {
+                        match_rule = match_rule.with_interface(value);
+                        Ok(())
```

Comment:
The `with_xxx` functions are quick and ergonomic; but panics if the path/interface/etc contains invalid characters. A parser could do better by constructing the path/interface first using a fallible function and return an error in case the path/interface/etc is invalid.

Reply:
I've mentioned this in the tests but ideally I'd like to implement TryInto here using the `dbus_validate_` family of functions.

Reply:
If you call `Path::new` or `Path::from_slice`, it will call into `dbus_validate_path` for you already. (And similar for `Interface` etc.)

There is no TryFrom/TryInto because nobody has implemented that and I wrote that code before TryFrom existed in std. Feel free to do so if you like.

Reply:
I just tried implementing it and failed due to the blanket implementation in core :( Basically it implements TryInto<U> for all types that implement Into<U> to help with interfaces. And From implicitly implements Into so we're out of luck unless we switch from From to TryFrom which would break a lot of usages but would eliminate the panic!'s.

Relevant issue: https://github.com/rust-lang/rust/issues/50133

Reply:
Oh, ok. Yeah switching from `From` to `TryFrom` is probably too backwards incompatible at this point...maybe for 0.10 if we ever release that.